### PR TITLE
gcc@6: Support for building on Linux

### PR DESCRIPTION
**Pull request by dholm**

### CODE REVIEW

**Comment by sjackman** [comment]
Hi, David. I would like to get `gcc@6` into Linuxbrew. I'm a bit busy this week, so I'm afraid I may not get to it real soon, but it's definitely on my radar. Here's two related PR that I opened for gcc@4.9 #1857 and gcc@5 #1860.

**Comment by sjackman** [comment]
TravisCI
```
g++-4.8 -std=gnu++98 -fno-PIE -c   -g -DIN_GCC     -fno-exceptions -fno-rtti -fasynchronous-unwind-tables -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wno-format -Wmissing-format-attribute -Woverloaded-virtual -pedantic -Wno-long-long -Wno-variadic-macros -Wno-overlength-strings -fno-common  -DHAVE_CONFIG_H -I. -I. -I../../gcc -I../../gcc/. -I../../gcc/../include -I../../gcc/../libcpp/include -I/home/linuxbrew/.linuxbrew/opt/gmp/include -I/home/linuxbrew/.linuxbrew/opt/mpfr/include -I/home/linuxbrew/.linuxbrew/opt/libmpc/include  -I../../gcc/../libdecnumber -I../../gcc/../libdecnumber/bid -I../libdecnumber -I../../gcc/../libbacktrace -I/home/linuxbrew/.linuxbrew/opt/isl/include  -o lto-compress.o -MT lto-compress.o -MMD -MP -MF ./.deps/lto-compress.TPo ../../gcc/lto-compress.c
../../gcc/lto-compress.c:34:18: fatal error: zlib.h: No such file or directory
 #include <zlib.h>
```
https://travis-ci.org/Linuxbrew/homebrew-core/builds/232322944#L1179

Add…
```ruby
depends_on "zlib" unless OS.mac?
```

**Comment by dholm** [comment]
@sjackman Done!

**Comment by sjackman** [comment]
```
/tmp/gcc@6-20170516-14694-1txask2/gcc-6.3.0/build/./gcc/xgcc -B/tmp/gcc@6-20170516-14694-1txask2/gcc-6.3.0/build/./gcc/ -B/home/linuxbrew/.linuxbrew/Cellar/gcc@6/6.3.0/x86_64-pc-linux-gnu/bin/ -B/home/linuxbrew/.linuxbrew/Cellar/gcc@6/6.3.0/x86_64-pc-linux-gnu/lib/ -isystem /home/linuxbrew/.linuxbrew/Cellar/gcc@6/6.3.0/x86_64-pc-linux-gnu/include -isystem /home/linuxbrew/.linuxbrew/Cellar/gcc@6/6.3.0/x86_64-pc-linux-gnu/sys-include    -g -O2 -m32 -O2  -g -O2 -DIN_GCC    -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wno-format -Wstrict-prototypes -Wmissing-prototypes -Wold-style-definition  -isystem ./include   -fpic -mlong-double-80 -DUSE_ELF_SYMVER -g -DIN_LIBGCC2 -fbuilding-libgcc -fno-stack-protector   -fpic -mlong-double-80 -DUSE_ELF_SYMVER -I. -I. -I../../.././gcc -I../../../../libgcc -I../../../../libgcc/. -I../../../../libgcc/../gcc -I../../../../libgcc/../include  -DHAVE_CC_TLS  -DUSE_TLS -o _gcov_merge_add.o -MT _gcov_merge_add.o -MD -MP -MF _gcov_merge_add.dep -DL_gcov_merge_add -c ../../../../libgcc/libgcov-merge.c
In file included from /usr/include/stdio.h:27:0,
                 from ../../../../libgcc/../gcc/tsystem.h:87,
                 from ../../../../libgcc/libgcov.h:41,
                 from ../../../../libgcc/libgcov-merge.c:26:
/usr/include/features.h:374:25: fatal error: sys/cdefs.h: No such file or directory
```
https://travis-ci.org/Linuxbrew/homebrew-core/builds/232749900#L1202

I'm not sure what's up with this error.

**Comment by dholm** [comment]
@sjackman On Ubuntu you must install `g++-multilib` since the non-multilib gcc/g++ will look for the header in `/usr/include` and not `/usr/include/x86_64-linux-gnu`.

**Comment by sjackman** [comment]
My understanding is that x86 Debian systems are all multilib (though you may or may not have the 32-bit headers/libraries installed), and on a 64-bit system it always searches `/usr/include/x86_64-linux-gnu` before `/usr/include`.

To test it out, you can modify the `circle.yml` in this same PR to install `g++-multilib`.

**Comment by dholm** [comment]
@sjackman Cool, pushed just now and changed `g++` to `g++-multilib`.

**Comment by dholm** [comment]
Weird, now it's complaining about zlib.h again even though I have it as a dependency.

**Comment by sjackman** [comment]
My best guess is to add
```ruby
"--with-zlib=#{Formula["zlib"].opt_prefix}",
```
or
```ruby
*("--with-system-zlib" if OS.mac?),
```
https://github.com/Linuxbrew/homebrew-core/blob/master/Formula/gcc@6.rb#L90

**Comment by dholm** [comment]
@sjackman Ah, right, I totally missed that!  Pushing it now.

File: Formula/gcc@6.rb [hunk 1]
```diff
@@ -77,17 +78,17 @@ def install
     arch = MacOS.prefer_64_bit? ? "x86_64" : "i686"
 
     args = [
-      "--build=#{arch}-apple-darwin#{osmajor}",
       "--prefix=#{prefix}",
-      "--libdir=#{lib}/gcc/#{version_suffix}",
+      ("--build=#{arch}-apple-darwin#{osmajor}" if OS.mac?),
```
Comment (sjackman) [inline Formula/gcc@6.rb]: Please keep this in the same order as on Homebrew, `build`, `prefix` then `libdir`.

File: Formula/gcc@6.rb [hunk 2]
```diff
@@ -77,17 +78,17 @@ def install
...
+      ("--libdir=#{lib}/gcc/#{version_suffix}" if OS.mac?),
       "--enable-languages=#{languages.join(",")}",
       # Make most executables versioned to avoid conflicts.
       "--program-suffix=-#{version_suffix}",
       "--with-gmp=#{Formula["gmp"].opt_prefix}",
       "--with-mpfr=#{Formula["mpfr"].opt_prefix}",
       "--with-mpc=#{Formula["libmpc"].opt_prefix}",
       "--with-isl=#{Formula["isl"].opt_prefix}",
-      "--with-system-zlib",
+      *("--with-system-zlib" if OS.mac?),
```
Comment (sjackman) [inline Formula/gcc@6.rb]: Instead of this chunk, have you tried
```ruby
"--with-zlib=#{Formula["zlib"].opt_prefix}"
```
?

**Comment by sjackman** [comment]
```
checking dynamic linker characteristics... checking whether strnlen is declared... yes
checking for getexecname... configure: error: Link tests are not allowed after GCC_NO_EXECUTABLES.
no
checking whether tests can run... make[2]: *** [configure-stage1-zlib] Error 1
```
https://travis-ci.org/Linuxbrew/homebrew-core/builds/238358418#L1195
I haven't seen that error before. It failed early, after 6 minutes.

File: Formula/gcc@6.rb (hunk 1 above)
Comment (dholm) [inline Formula/gcc@6.rb]: Fixed

File: Formula/gcc@6.rb (hunk 2 above)
Comment (dholm) [inline Formula/gcc@6.rb]: Testing it now..

**Comment by sjackman** [comment]
```
==> make bootstrap
......................................................................................................
command umask 022; PATH="/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:$PATH"; eval "$(perl -I$HOME/perl5/lib/perl5 -Mlocal::lib)"; brew install patchelf && brew tap homebrew/science && brew tap linuxbrew/xorg && brew test-bot --tap=homebrew/core --keep-old took more than 120 minutes to run
```
https://circleci.com/gh/Linuxbrew/homebrew-core/5943

That looks like progress.

Try increasing the number of threads on CircleCI. Too many and you'll exceed the 4 GB memory limit. The default is 8. You can increase it up to 32.

```ruby
# Increase the number of threads to avoid timing out.
ENV["MAKEFLAGS"] = "-j12" if ENV["CIRCLECI"]
```

**Comment by rohituks** [comment]
`==> ../configure --build=x86_64-apple-darwin2.6.32-642.15.1.el6.x86_64 --prefix=/home/cadappl/cadappl_linux_i386/linuxbrew/Cellar/gcc@6/6.3.0 --libdir=/home/cadappl/cadappl_linux_i386/linuxbrew/Cellar/gcc@6/6.3.0/lib/gcc/6 --enable-langu
Last 15 lines from /home/supportg4id/.cache/Homebrew/Logs/gcc@6/01.configure:
--with-isl=/home/cadappl/cadappl_linux_i386/linuxbrew/opt/isl
--with-system-zlib
--enable-stage1-checking
--enable-checking=release
--enable-lto
--with-build-config=bootstrap-debug
--disable-werror
--with-pkgversion=Homebrew GCC 6.3.0
--with-bugurl=https://github.com/Homebrew/homebrew-core/issues
--with-dwarf2
--disable-nls
--enable-multilib

checking build system type... Invalid configuration `x86_64-apple-darwin2.6.32-642.15.1.el6.x86_64': machine `x86_64-apple-darwin2.6.32' not recognized
configure: error: /bin/sh ../config.sub x86_64-apple-darwin2.6.32-642.15.1.el6.x86_64 failed
`

Unable to install gcc@6 in linux redhat cluster environment

**Comment by sjackman** [comment]
Please enclose copy-and-paste blocks in triple back ticks.
See https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code

**Comment by sjackman** [comment]
@rohituks `gcc@6` does not yet work on Linuxbrew. This PR is attempting to fix `gcc@6` for Linuxbrew. You can install GCC 5 with `brew install gcc` if you like.

**Comment by rohituks** [comment]
Thanks. I have gcc@5 working fine, but i got this option today to install gcc@6 in linuxbrew which was not visible till last week.
So, tried installing that and faced this issue

**Comment by sjackman** [comment]
I recently had success building `gcc@4.9` for Linuxbrew. I'll next take a look at getting `gcc@6` working.

**Comment by dholm** [comment]
@sjackman Pushing change with threading increased to 16..

File: Formula/gcc@6.rb [hunk 3]
```diff
@@ -77,17 +81,17 @@ def install
...
+      ("--libdir=#{lib}/gcc/#{version_suffix}" if OS.mac?),
       "--enable-languages=#{languages.join(",")}",
       # Make most executables versioned to avoid conflicts.
       "--program-suffix=-#{version_suffix}",
       "--with-gmp=#{Formula["gmp"].opt_prefix}",
       "--with-mpfr=#{Formula["mpfr"].opt_prefix}",
       "--with-mpc=#{Formula["libmpc"].opt_prefix}",
       "--with-isl=#{Formula["isl"].opt_prefix}",
-      "--with-system-zlib",
+      "--with-zlib=#{Formula["zlib"].opt_prefix}",
```
Comment (sjackman) [inline Formula/gcc@6.rb]: Try instead leaving `--with-system-zlib` as is and add:
``ruby
    ENV["CPPFLAGS"] = "-I#{Formula["zlib"].include}" unless OS.mac?
```
This fix worked for `gcc@4.9`.

Comment (dholm) [inline Formula/gcc@6.rb]: Done!

**Comment by sjackman** [comment]
```
/tmp/gcc@6-20170608-14221-1ssthfb/gcc-6.3.0/build/./gcc/xgcc -B/tmp/gcc@6-20170608-14221-1ssthfb/gcc-6.3.0/build/./gcc/ -B/home/linuxbrew/.linuxbrew/Cellar/gcc@6/6.3.0/x86_64-pc-linux-gnu/bin/ -B/home/linuxbrew/.linuxbrew/Cellar/gcc@6/6.3.0/x86_64-pc-linux-gnu/lib/ -isystem /home/linuxbrew/.linuxbrew/Cellar/gcc@6/6.3.0/x86_64-pc-linux-gnu/include -isystem /home/linuxbrew/.linuxbrew/Cellar/gcc@6/6.3.0/x86_64-pc-linux-gnu/sys-include    -g -O2 -m32 -O2  -g -O2 -DIN_GCC    -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wno-format -Wstrict-prototypes -Wmissing-prototypes -Wold-style-definition  -isystem ./include   -fpic -mlong-double-80 -DUSE_ELF_SYMVER -g -DIN_LIBGCC2 -fbuilding-libgcc -fno-stack-protector   -fpic -mlong-double-80 -DUSE_ELF_SYMVER -I. -I. -I../../.././gcc -I../../../../libgcc -I../../../../libgcc/. -I../../../../libgcc/../gcc -I../../../../libgcc/../include  -DHAVE_CC_TLS  -DUSE_TLS -o _gcov_merge_add.o -MT _gcov_merge_add.o -MD -MP -MF _gcov_merge_add.dep -DL_gcov_merge_add -c ../../../../libgcc/libgcov-merge.c
In file included from /usr/include/stdio.h:27:0,
                 from ../../../../libgcc/../gcc/tsystem.h:87,
                 from ../../../../libgcc/libgcov.h:41,
                 from ../../../../libgcc/libgcov-merge.c:26:
/usr/include/features.h:374:25: fatal error: sys/cdefs.h: No such file or directory
 #  include <sys/cdefs.h>
                         ^
compilation terminated.
make[5]: *** [_gcov_merge_add.o] Error 1
make[5]: Leaving directory `/tmp/gcc@6-20170608-14221-1ssthfb/gcc-6.3.0/build/x86_64-pc-linux-gnu/32/libgcc'
```
Try adding
```ruby
  # enabling multilib on a host that can't run 64-bit results in build failures
  if OS.mac?
    option "without-multilib", "Build without multilib support" if MacOS.prefer_64_bit?
  else
    option "with-multilib", "Build with multilib support"
  end
```
See the `gcc` formula for an example.

**Comment by sjackman** [comment]
CircleCi `took more than 120 minutes to run` with 16 threads. Try increasing the number of threads to 24.

**Comment by dholm** [comment]
@sjackman Pushing now..

**Comment by sjackman** [comment]
With 32 processes…
> Your build has exceeded the memory limit of 4G on 1 container.

https://circleci.com/gh/Linuxbrew/homebrew-core/6069
https://6069-56942570-gh.circle-artifacts.com/0/tmp/memory-usage.txt

Try 24 processes.

**Comment by dholm** [comment]
@sjackman Reduced to 24.

**Pull request closed without merging**